Firestore/core/test/firebase/firestore/local/mutation_queue_test.cc

```diff
@@ -0,0 +1,432 @@
+#include "Firestore/core/src/firebase/firestore/model/document_key.h"
+#include "Firestore/core/src/firebase/firestore/model/document_key_set.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation.h"
+#include "Firestore/core/src/firebase/firestore/model/mutation_batch.h"
+#include "Firestore/core/src/firebase/firestore/model/set_mutation.h"
+#include "Firestore/core/test/firebase/firestore/testutil/testutil.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+using auth::User;
+using model::DocumentKey;
+using model::DocumentKeySet;
+using model::kBatchIdUnknown;
+using model::Mutation;
+using model::MutationBatch;
+using model::SetMutation;
+using nanopb::ByteString;
+using testutil::Key;
+using testutil::Map;
+using testutil::Query;
+
+MutationQueueTestBase::MutationQueueTestBase(
+    std::unique_ptr<Persistence> persistence)
+    : persistence_(std::move(persistence)),
+      mutation_queue_(persistence_->GetMutationQueueForUser(User("user"))) {
+}
+
+MutationQueueTestBase::~MutationQueueTestBase() = default;
+
+/**
+ * Creates a new MutationBatch with the given key, the next batch ID and a set
+ * of dummy mutations.
+ */
+MutationBatch MutationQueueTestBase::AddMutationBatch(const std::string& key) {
+  SetMutation mutation = testutil::SetMutation(key, Map("a", 1));
+
+  MutationBatch batch =
+      mutation_queue_->AddMutationBatch(Timestamp::Now(), {}, {mutation});
+  return batch;
+}
+
+/**
+ * Creates an array of batches containing @a number dummy MutationBatches. Each
+ * has a different batch_id.
+ */
+std::vector<MutationBatch> MutationQueueTestBase::CreateBatches(int number) {
+  std::vector<MutationBatch> batches;
+
+  for (int i = 0; i < number; i++) {
+    MutationBatch batch = AddMutationBatch();
+    batches.push_back(batch);
+  }
+
+  return batches;
+}
+
+/** Returns the number of mutation batches in the mutation queue. */
+size_t MutationQueueTestBase::BatchCount() {
+  return mutation_queue_->AllMutationBatches().size();
+}
+
+/**
+ * Removes the first n entries from the the given batches and returns them.
+ *
+ * @param n The number of batches to remove.
+ * @param batches The array to mutate, removing entries from it.
+ * @return A new array containing all the entries that were removed from @a
+ * batches.
+ */
+std::vector<MutationBatch> MutationQueueTestBase::RemoveFirstBatches(
+    size_t n, std::vector<MutationBatch>* batches) {
+  std::vector<MutationBatch> removed(batches->begin(), batches->begin() + n);
+  batches->erase(batches->begin(), batches->begin() + n);
+
+  for (const MutationBatch& batch : removed) {
+    mutation_queue_->RemoveMutationBatch(batch);
+  }
+  return removed;
+}
+
+MutationQueueTest::MutationQueueTest() : MutationQueueTestBase(GetParam()()) {
+}
+
+TEST_P(MutationQueueTest, CountBatches) {
+  persistence_->Run("test_count_batches", [&] {
+    ASSERT_EQ(0, BatchCount());
```

Comment:
I think we generally tend to use `EXPECT` instead of `ASSERT` in our tests. I'm not against it, just curious about the rationale.

Reply:
`EXPECT` makes sense when the assertions are independent. In this test, if the prior state is wrong the next assertions are also going to be broken. `ASSERT` short circuits once the rest of the test will become invalid.

I've changed a few to `EXPECT` below where there are no assertion dependencies.